File: code/TemplateStudioForUWP/Templates/Uwp/Services/IdentityOptionalLogin/ViewModels/SettingsViewModel_postaction.cs
//{[{
using Param_RootNamespace.Core.Helpers;
using Param_RootNamespace.Core.Services;
//}]}
namespace Param_RootNamespace.ViewModels
{
    public class SettingsViewModel : System.ComponentModel.INotifyPropertyChanged
    {
//{[{
        private UserDataService UserDataService => Singleton<UserDataService>.Instance;

        private IdentityService IdentityService => Singleton<IdentityService>.Instance;

//}]}
        private ElementTheme _elementTheme = ThemeSelectorService.Theme;
//{[{
        private RelayCommand _logInCommand;
        private RelayCommand _logOutCommand;
        private bool _isLoggedIn;
        private bool _isBusy;
        private UserViewModel _user;
//}]}
        public ElementTheme ElementTheme
        {
        }
//^^
//{[{
        public RelayCommand LogInCommand => _logInCommand ?? (_logInCommand = new RelayCommand(OnLogIn, () => !IsBusy));

        public RelayCommand LogOutCommand => _logOutCommand ?? (_logOutCommand = new RelayCommand(OnLogOut, () => !IsBusy));

        public bool IsLoggedIn
        {
            get { return _isLoggedIn; }
            set { Param_Setter(ref _isLoggedIn, value); }
        }

        public bool IsBusy
        {
            get => _isBusy;
            set
            {
                Param_Setter(ref _isBusy, value);
            }
        }

        public UserViewModel User
        {
            get { return _user; }
            set { Param_Setter(ref _user, value); }
        }
//}]}

        public SettingsViewModel()
        {
        }

        public async Task InitializeAsync()
        {
            VersionDescription = GetVersionDescription();
//^^
//{[{
            IdentityService.LoggedIn += OnLoggedIn;
            IdentityService.LoggedOut += OnLoggedOut;
            UserDataService.UserDataUpdated += OnUserDataUpdated;
            IsLoggedIn = IdentityService.IsLoggedIn();
            User = await UserDataService.GetUserAsync();
//}]}
        }

//^^
//{[{
        public void UnregisterEvents()
        {
            IdentityService.LoggedIn -= OnLoggedIn;
            IdentityService.LoggedOut -= OnLoggedOut;
            UserDataService.UserDataUpdated -= OnUserDataUpdated;
        }

        private void OnUserDataUpdated(object sender, UserViewModel userData)
        {
            User = userData;
        }

        private async void OnLogIn()
        {
            IsBusy = true;
            var loginResult = await IdentityService.LoginAsync();
            if (loginResult != LoginResultType.Success)
            {
                await AuthenticationHelper.ShowLoginErrorAsync(loginResult);
                IsBusy = false;
            }
        }

        private async void OnLogOut()
        {
            IsBusy = true;
            await IdentityService.LogoutAsync();
        }

        private void OnLoggedIn(object sender, EventArgs e)
        {
            IsLoggedIn = true;
            IsBusy = false;
        }

        private void OnLoggedOut(object sender, EventArgs e)
        {
            User = null;
            IsLoggedIn = false;
            IsBusy = false;
        }
//}]}
    }
}

File: code/TemplateStudioForUWP/Templates/Uwp/Services/IdentityOptionalLogin._VB/ViewModels/SettingsViewModel_postaction.vb
'{[{
Imports Param_RootNamespace.Core.Helpers
Imports Param_RootNamespace.Core.Services
'}]}
Namespace ViewModels
    Public Class SettingsViewModel
        Inherits System.ComponentModel.INotifyPropertyChanged

'{[{

        Private ReadOnly Property UserDataService As UserDataService
            Get
                Return Singleton(Of UserDataService).Instance
            End Get
        End Property

        Private ReadOnly Property IdentityService As IdentityService
            Get
                Return Singleton(Of IdentityService).Instance
            End Get
        End Property
'}]}

        Private _elementTheme As ElementTheme = ThemeSelectorService.Theme
'{[{
        Private _logInCommand As RelayCommand
        Private _logOutCommand As RelayCommand
        Private _isLoggedIn As Boolean
        Private _isBusy As Boolean
        Private _user As UserViewModel
'}]}
        Public Property ElementTheme As ElementTheme

'^^
'{[{
        Public ReadOnly Property LogInCommand As RelayCommand
            Get
                If _loginCommand Is Nothing Then
                    _loginCommand = New RelayCommand(AddressOf OnLogin, Function() Not IsBusy)
                End If

                Return _loginCommand
            End Get
        End Property

        Public ReadOnly Property LogOutCommand As RelayCommand
            Get
                If _logOutCommand Is Nothing Then
                    _logOutCommand = New RelayCommand(AddressOf OnLogOut, Function() Not IsBusy)
                End If

                Return _logOutCommand
            End Get
        End Property

        Public Property IsLoggedIn As Boolean
            Get
                Return _isLoggedIn
            End Get
            Set(value As Boolean)
                [Param_Setter](_isLoggedIn, value)
            End Set
        End Property

        Public Property IsBusy As Boolean
            Get
                Return _isBusy
            End Get
            Set(value As Boolean)
                [Param_Setter](_isBusy, value)
            End Set
        End Property

        Public Property User As UserViewModel
            Get
                Return _user
            End Get
            Set(value As UserViewModel)
                [Param_Setter](_user, value)
            End Set
        End Property

'}]}

        Public Sub New()
        End Sub

        Public Async Function InitializeAsync() As Task
            VersionDescription = GetVersionDescription()
'^^
'{[{
            AddHandler IdentityService.LoggedIn, AddressOf OnLoggedIn
            AddHandler IdentityService.LoggedOut, AddressOf OnLoggedOut
            AddHandler UserDataService.UserDataUpdated, AddressOf OnUserDataUpdated
            IsLoggedIn = IdentityService.IsLoggedIn()
            User = Await UserDataService.GetUserAsync()
'}]}
        End Function

'^^
'{[{
        Public Sub UnregisterEvents()
            RemoveHandler IdentityService.LoggedIn, AddressOf OnLoggedIn
            RemoveHandler IdentityService.LoggedOut, AddressOf OnLoggedOut
            RemoveHandler UserDataService.UserDataUpdated, AddressOf OnUserDataUpdated
        End Sub

        Private Sub OnUserDataUpdated(sender As Object, userData As UserViewModel)
            User = userData
        End Sub

        Private Async Sub OnLogIn()
            IsBusy = True
            Dim loginResult = Await IdentityService.LoginAsync()

            If loginResult <> LoginResultType.Success Then
                Await AuthenticationHelper.ShowLoginErrorAsync(loginResult)
                IsBusy = False
            End If
        End Sub

        Private Async Sub OnLogOut()
            IsBusy = True
            Await IdentityService.LogoutAsync()
        End Sub

        Private Sub OnLoggedIn(sender As Object, e As EventArgs)
            IsLoggedIn = True
            IsBusy = False
        End Sub

        Private Sub OnLoggedOut(sender As Object, e As EventArgs)
            User = Nothing
            IsLoggedIn = False
            IsBusy = False
        End Sub
'}]}
    End Class
End Namespace
